pkg/controller/vmi/controller.go

```diff
@@ -0,0 +1,109 @@
+package vmi
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+
+	lbv1 "github.com/harvester/harvester-load-balancer/pkg/apis/loadbalancer.harvesterhci.io/v1alpha1"
+	"github.com/harvester/harvester-load-balancer/pkg/config"
+	ctllbv1 "github.com/harvester/harvester-load-balancer/pkg/generated/controllers/loadbalancer.harvesterhci.io/v1alpha1"
+	lbpkg "github.com/harvester/harvester-load-balancer/pkg/lb"
+	"github.com/harvester/harvester-load-balancer/pkg/lb/servicelb"
+	"github.com/harvester/harvester-load-balancer/pkg/utils"
+)
+
+const controllerName = "harvester-lb-vmi-controller"
+
+type Handler struct {
+	lbClient ctllbv1.LoadBalancerClient
+	lbCache  ctllbv1.LoadBalancerCache
+
+	lbManager lbpkg.Manager
+}
+
+func Register(ctx context.Context, management *config.Management) error {
+	vmis := management.KubevirtFactory.Kubevirt().V1().VirtualMachineInstance()
+	lbs := management.LbFactory.Loadbalancer().V1alpha1().LoadBalancer()
+
+	handler := &Handler{
+		lbClient: lbs,
+		lbCache:  lbs.Cache(),
+
+		lbManager: management.LBManager,
+	}
+
+	vmis.OnChange(ctx, controllerName, handler.OnChange)
+	vmis.OnRemove(ctx, controllerName, handler.OnRemove)
+
+	return nil
+}
+
+func (h *Handler) OnChange(key string, vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
+	if vmi == nil || vmi.DeletionTimestamp != nil {
+		return nil, nil
+	}
+	logrus.Infof("VMI %s has been changed", vmi.Name)
+
+	lbs, err := h.lbCache.List("", labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, lb := range lbs {
+		// skip the cluster LB
+		if lb.Spec.WorkloadType == lbv1.Cluster {
+			continue
+		}
+
+		selector, err := utils.NewSelector(lb.Spec.BackendServerSelector)
+		if err != nil {
+			return nil, err
+		}
+
+		// add or update the backend server to the matched load balancer
```

Comment:
vmi changes frequently; for most of the time, it is not related to lb
the lb checks vmi via iterating endpointslice,  each time

it seems need a fast way,  e.g. lb looks record info of vmi (e.g. id, ip), to have a one step match check